File: typescript/base/src/main/java/lang/Double.ts
import { IllegalArgumentException } from "./IllegalArgumentException.js";

/** Java double helpers, including lossless IEEE-754 bit conversions. */
export class Double {
    public static readonly MIN_VALUE = 4.9406564584124654e-324;
    public static readonly MIN_NORMAL = 2.2250738585072014e-308;
    public static readonly MAX_VALUE = 1.7976931348623157e308;
    public static readonly POSITIVE_INFINITY = Infinity;
    public static readonly NEGATIVE_INFINITY = -Infinity;
    public static readonly NaN = NaN;
    public static readonly SIZE = 64;
    public static readonly BYTES = 8;

    public static valueOf(value: number | string): number {
        return typeof value === "number" ? value : Double.parseDouble(value);
    }
    public static parseDouble(text: string): number {
        if (text.length === 0 || text.trim() !== text)
            throw new IllegalArgumentException(`For input string: "${text}"`);
        if (text === "NaN") return NaN;
        if (text === "Infinity" || text === "+Infinity") return Infinity;
        if (text === "-Infinity") return -Infinity;
        const value = Number(
            text.endsWith("d") || text.endsWith("D") || text.endsWith("f") || text.endsWith("F")
                ? text.slice(0, -1)
                : text,
        );
        if (Number.isNaN(value)) throw new IllegalArgumentException(`For input string: "${text}"`);
        return value;
    }
    public static isNaN(value: number): boolean {
        return Number.isNaN(value);
    }
    public static isInfinite(value: number): boolean {
        return value === Infinity || value === -Infinity;
    }
    public static isFinite(value: number): boolean {
        return Number.isFinite(value);
    }
    public static compare(left: number, right: number): number {
        if (left < right) return -1;
        if (left > right) return 1;
        const a = Double.doubleToLongBits(left);
        const b = Double.doubleToLongBits(right);
        return a === b ? 0 : a < b ? -1 : 1;
    }
    public static doubleToRawLongBits(value: number): bigint {
        const view = new DataView(new ArrayBuffer(8));
        view.setFloat64(0, value, false);
        return view.getBigInt64(0, false);
    }
    public static doubleToLongBits(value: number): bigint {
        return Number.isNaN(value) ? 0x7ff8_0000_0000_0000n : Double.doubleToRawLongBits(value);
    }
    public static longBitsToDouble(bits: bigint): number {
        const view = new DataView(new ArrayBuffer(8));
        view.setBigInt64(0, BigInt.asIntN(64, bits), false);
        return view.getFloat64(0, false);
    }
    public static hashCode(value: number): number {
        const bits = BigInt.asUintN(64, Double.doubleToLongBits(value));
        return Number((bits ^ (bits >> 32n)) & 0xffff_ffffn) | 0;
    }
    /**
     * Java's `Double.toString` layout: a plain decimal with at least one
     * fraction digit when the magnitude is in [1e-3, 1e7), and the
     * `<digit>.<digits>E<exponent>` computerized scientific form otherwise.
     * JavaScript instead drops the trailing `.0`, writes `1e-7` in lower case,
     * and keeps plain notation up to 1e21.
     *
     * The significant digits are the shortest ones that round-trip, which is
     * what both runtimes compute; the pre-19 Java implementation emits one
     * extra digit for a small set of values, and that residual difference is
     * not reproduced here.
     */
    public static toString(value: number): string {
        if (Number.isNaN(value)) return "NaN";
        if (value === Infinity) return "Infinity";
        if (value === -Infinity) return "-Infinity";
        if (value === 0) return Object.is(value, -0) ? "-0.0" : "0.0";

        const sign = value < 0 ? "-" : "";
        const magnitude = Math.abs(value);
        const exponential = magnitude.toExponential();
        const [mantissa, exponentText] = exponential.split("e");
        const exponent = Number(exponentText);
        const digits = mantissa!.replace(".", "");

        if (magnitude >= 1e-3 && magnitude < 1e7) {
            if (exponent >= 0) {
                const integerPart = digits.slice(0, exponent + 1).padEnd(exponent + 1, "0");
                const fractionPart = digits.slice(exponent + 1);
                return `${sign}${integerPart}.${fractionPart === "" ? "0" : fractionPart}`;
            }
            return `${sign}0.${"0".repeat(-exponent - 1)}${digits}`;
        }
        const fractionPart = digits.slice(1);
        return `${sign}${digits.charAt(0)}.${fractionPart === "" ? "0" : fractionPart}E${exponent}`;
    }
}

File: cpp/base/src/main/java/lang/Double.cpp
#include <cerrno>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <cstring>

#include "java/lang/Double.h"
#include "java/lang/Integer.h"
#include "java/lang/NumberFormatException.h"

namespace java {

namespace {
bool isSpace(char c)
{
    return c == ' ' || c == '\t' || c == '\n' || c == '\r' || c == '\f';
}
}

double Double::parseDouble(const java::String& text)
{
    const char* start = text.c_str();
    while ( isSpace(*start) ) {
        start++;
    }
    if ( *start == '\0' ) {
        throw NumberFormatException(
            text.isEmpty() ? java::String("empty String") : text);
    }
    char* end = nullptr;
    double value = std::strtod(start, &end);
    if ( end == start ) {
        throw NumberFormatException(
            java::String("For input string: \"") + text + "\"");
    }
    // Java accepts an optional float/double suffix
    if ( *end == 'd' || *end == 'D' || *end == 'f' || *end == 'F' ) {
        end++;
    }
    while ( isSpace(*end) ) {
        end++;
    }
    if ( *end != '\0' ) {
        throw NumberFormatException(
            java::String("For input string: \"") + text + "\"");
    }
    return value;
}

java::String Double::toString(double value)
{
    if ( std::isnan(value) ) {
        return "NaN";
    }
    if ( std::isinf(value) ) {
        return value > 0 ? "Infinity" : "-Infinity";
    }
    if ( value == 0.0 ) {
        return std::signbit(value) ? "-0.0" : "0.0";
    }

    // Shortest mantissa that reads back as the same number
    char buffer[64];
    int precision;
    for ( precision = 1; precision <= 17; precision++ ) {
        std::snprintf(buffer, sizeof(buffer), "%.*e", precision - 1, value);
        if ( std::strtod(buffer, nullptr) == value ) {
            break;
        }
    }

    // buffer is "[-]d.ddde[+-]xx": split digits and exponent
    bool negative = buffer[0] == '-';
    const char* p = negative ? buffer + 1 : buffer;
    char digits[32];
    int digitCount = 0;
    while ( *p != 'e' && *p != '\0' ) {
        if ( *p != '.' ) {
            digits[digitCount++] = *p;
        }
        p++;
    }
    digits[digitCount] = '\0';
    int exponent = (*p == 'e') ? std::atoi(p + 1) : 0;
    while ( digitCount > 1 && digits[digitCount - 1] == '0' ) {
        digits[--digitCount] = '\0';
    }

    java::String out = negative ? "-" : "";
    char one[2] = {'\0', '\0'};
    double magnitude = std::fabs(value);

    if ( magnitude >= 1e-3 && magnitude < 1e7 ) {
        if ( exponent >= 0 ) {
            int i;
            for ( i = 0; i <= exponent; i++ ) {
                one[0] = i < digitCount ? digits[i] : '0';
                out += one;
            }
            out += ".";
            if ( exponent + 1 >= digitCount ) {
                out += "0";
            }
            else {
                out += digits + exponent + 1;
            }
        }
        else {
            out += "0.";
            int i;
            for ( i = 0; i < -exponent - 1; i++ ) {
                out += "0";
            }
            out += digits;
        }
        return out;
    }

    one[0] = digits[0];
    out += one;
    out += ".";
    out += digitCount > 1 ? java::String(digits + 1) : java::String("0");
    out += "E";
    out += java::String::valueOf(exponent);
    return out;
}

int Integer::parseInt(const java::String& text)
{
    const char* start = text.c_str();
    if ( *start == '\0' ) {
        throw NumberFormatException(
            java::String("For input string: \"") + text + "\"");
    }
    const char* p = start;
    if ( *p == '+' || *p == '-' ) {
        p++;
    }
    if ( *p == '\0' ) {
        throw NumberFormatException(
            java::String("For input string: \"") + text + "\"");
    }
    for ( const char* q = p; *q != '\0'; q++ ) {
        if ( *q < '0' || *q > '9' ) {
            throw NumberFormatException(
                java::String("For input string: \"") + text + "\"");
        }
    }
    errno = 0;
    long long value = std::strtoll(start, nullptr, 10);
    if ( errno != 0 || value < MIN_VALUE || value > MAX_VALUE ) {
        throw NumberFormatException(
            java::String("For input string: \"") + text + "\"");
    }
    return (int)value;
}

}
